makefiles/arch/riscv.inc.mk

```
@@ -30,8 +30,15 @@ TARGET_ARCH_RISCV ?= \
 TARGET_ARCH ?= $(TARGET_ARCH_RISCV)
 
 # define build specific options
-CFLAGS_CPU   = -march=rv32imac -mabi=ilp32 -mcmodel=medlow -msmall-data-limit=8
-CFLAGS_LINK  = -nostartfiles -ffunction-sections -fdata-sections
+CFLAGS_CPU   = -march=rv32imac -mabi=ilp32
+ifeq ($(TOOLCHAIN),llvm)
+  # Always use riscv32-none-elf as target triple for clang, as some
+  # autodetected gcc target triples are incompatible with clang
+  TARGET_ARCH_LLVM := riscv32-none-elf
```

```suggestion
  TARGET_ARCH_LLVM ?= riscv32-none-elf
```

With this, you can override this value with a different triple.

---

This flexibility is luckily not needed for LLVM. Technically, `riscv32-none-elf` is the correct target triplet for 32 bit embedded (bare metal) RISC-V targets. Differently to GCC, the target triplet is just an argument rather than a prefix to the binaries and managed upstream, so unless one manually patches clang to only work with incorrect target triples, nobody will need to change this.

---

I was thinking of the riscv64-none-elf triplet available in the Debian/Ubuntu package manager. On Debian/Ubuntu, that would need picolibc but IIUC, it might be possible to use llvm also with picolibc ?

---

> I was thinking of the riscv64-none-elf triplet available in the Debian/Ubuntu package manager.

No, the libraries are resolved using the triple in `TARGET_ARCH`, which can be distinct from `TARGET_ARCH_LLVM`. The issue is that `riscv64-none-elf` is the correct triple for the 64 bit RISC-V, but we only support 32 bit so far. A `riscv64-none-elf` multilib GCC toolchain would support both 32 and 64 bit flavors, but `clang` wants the correct triple.

E.g. when the c lib is installed as `riscv-none-embed`, this is a completely invalid target triple. But splitting the target triple used to resolve the GCC toolchain and c library from the target triple passed to LLVM, we overcome the issues.

> IIUC, it might be possible to use llvm also with picolibc ?

Yes, starting with `clang11` it is possible. `clang10` doesn't understand the `-specs=` parameter.

---

Thanks for the clarifications @maribu. Makes sense to me now.